database/postgres/postgres.go

```diff
@@ -332,20 +332,19 @@ func (p *Postgres) Drop() error {
 	return nil
 }
 
-func (p *Postgres) ensureVersionTable() error {
-	// check if migration table exists
-	var count int
-	query := `SELECT COUNT(1) FROM information_schema.tables WHERE table_name = $1 AND table_schema = (SELECT current_schema()) LIMIT 1`
-	if err := p.conn.QueryRowContext(context.Background(), query, p.config.MigrationsTable).Scan(&count); err != nil {
-		return &database.Error{OrigErr: err, Query: []byte(query)}
-	}
-	if count == 1 {
-		return nil
+func (p *Postgres) ensureVersionTable() (err error) {
+	if err = p.Lock(); err != nil {
+		return err
 	}
 
-	// if not, create the empty migration table
-	query = `CREATE TABLE "` + p.config.MigrationsTable + `" (version bigint not null primary key, dirty boolean not null)`
-	if _, err := p.conn.ExecContext(context.Background(), query); err != nil {
+	defer func() {
+		if e := p.Unlock(); err == nil {
+			err = e
```

Comment:
The error wasn't being clobbered since it was only being set when the outer error was nil. But we were loosing the information of this error when the outer error was not nil. So wrapping is beneficial in this case. Any specific approach you prefer?

Reply:
At a quick glance, it seems like "error values" topic is most applicable.

There are some [implementations of the draft](https://github.com/golang/go/wiki/Go2ErrorValuesFeedback#draft-implementation). Of those, I'd lean towards `golang.org/x/exp/errors` but I'm not feeling particularly bold...

Reply:
Marking as resolved based on my last commit. Let me know if it looks ok to you.

Reply:
I have experience with https://github.com/pkg/errors and don't think we should rely on golang.org/x/exp/errors which is experimental at this stage. I'll make a change to use https://github.com/pkg/errors

Reply:
Looking at it better, these two errors are more like a multi error rather than a chain, since they both happen in the same stack frame. So I used https://github.com/hashicorp/go-multierror

Reply:
Using `hashicorp/go-multierror` sounds good.

You'll need to use [dep](https://github.com/golang/dep) to add a dependency on `github.com/hashicorp/go-multierror` for Go 1.10.x support. e.g. `dep ensure -add github.com/hashicorp/go-multierror@v1.0.0`

Reply:
Looking at this project's root directory, I see a `go.mod` and `go.sum`. Are we using `dep` as well? I'm confused.

Reply:
> I'm confused.

Rightfully so!

I *believe* that Go 1.10.x does not support Go modules. Only Go 1.11+ does. Go 1.10.3 added support for interpreting `/vN` imports, but doesn't use `go.mod`, so we still need to use `dep` to manage dependencies for Go 1.10.x
See `.travis.yml` on how the split is done.

I have a test branch that removes `dep`, but the [Go 1.10.x run fails](https://travis-ci.org/dhui/migrate/builds/479586545) b/c `go.mod` is ignored.

Please correct me if I'm wrong or missing something.

Reply:
Makes sense to me. I'll fix, thanks for the explanation!